File: Detect cycle in a directed graph - GFG/detect-cycle-in-a-directed-graph.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class DriverClass {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            int V = sc.nextInt();
            int E = sc.nextInt();
            for (int i = 0; i < V; i++)
                list.add(i, new ArrayList<Integer>());
            for (int i = 0; i < E; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                list.get(u).add(v);
            }
            if (new Solution().isCyclic(V, list) == true)
                System.out.println("1");
            else
                System.out.println("0");
        }
    }
}
// } Driver Code Ends


/*Complete the function below*/

class Solution {
    public boolean isCyclic(int v, ArrayList<ArrayList<Integer>> adj) {
        boolean visited[] = new boolean[v];
        boolean callStack[] = new boolean[v];
        for(int i=0;i<v;i++){
            if(!visited[i]){
                if(dfs(i, adj, visited, callStack)){
                    return true;
                }
            }
        }
        return false;
    }
    public boolean dfs(int v, ArrayList<ArrayList<Integer>> adj, boolean visited[], boolean callStack[]){
        visited[v]=true;
        callStack[v]=true;
        ArrayList<Integer> neighbors = adj.get(v);
        for(int neighbor: neighbors){
            if(!visited[neighbor]){
                if(dfs(neighbor, adj, visited, callStack)){
                    return true;
                }
            }
            else if(callStack[neighbor]){
               return true;
            }
        }
        callStack[v]=false;
        return false;
    }
}

File: Detect cycle in a directed graph - GFG/detect-cycle-in-a-directed-graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  public:
    bool dfs(int v, vector<int>* adj, vector<bool>& visited, vector<bool>& callStack) {
        visited[v] = true;
        callStack[v] = true;
        
        vector<int>& neighbors = adj[v];
        for (int neighbor : neighbors) {
            if (!visited[neighbor]) {
                if (dfs(neighbor, adj, visited, callStack)) {
                    return true;
                }
            } else if (callStack[neighbor]) {
                return true;
            }
        }
        
        callStack[v] = false;
        return false;
    }
    bool isCyclic(int v, vector<int> adj[]) {
        vector<bool> visited(v, false);
        vector<bool> callStack(v, false);
        
        for (int i = 0; i < v; i++) {
            if (!visited[i]) {
                if (dfs(i, adj, visited, callStack)) {
                    return true;
                }
            }
        }
        
        return false;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {
        int V, E;
        cin >> V >> E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
        }

        Solution obj;
        cout << obj.isCyclic(V, adj) << "\n";
    }

    return 0;
}

// } Driver Code Ends